template/web/app/paymaster-bundler/_utils/fetchNFTs.ts

```diff
@@ -0,0 +1,60 @@
+import { SmartAccountClient } from 'permissionless';
+import { PublicClient } from 'viem';
+
+import { nftAbi } from '../_components/abi';
+import { NFTType } from '../types';
+
+export default async function fetchNFTs(smartAccount: SmartAccountClient, client: PublicClient) {
+  // Get # of NFTs owned by address
+  const address = smartAccount.account?.address;
+
+  const numTokens = await client.readContract({
+    address: '0x66519FCAee1Ed65bc9e0aCc25cCD900668D3eD49',
+    abi: nftAbi,
+    functionName: 'balanceOf',
+    args: [address],
+  });
+
+  // Get the token IDs and metadata by token ID
+  var tokens = [] as string[];
+
+  for (let i = 0; i < Number(numTokens); i++) {
+    const tokenID = await client.readContract({
```

Comment:
This is crazy inefficient for users who own a large amount of tokens, it will end up locking up the UI and sending a large amount of requests to the RPC provider. I know this is not a simple solution, however, we should probably use a data aggregator here. cc: @Zizzamia .. WDYT about integrating with an API vendor for these type of aggregated data lookups?

Reply:
keep in mind that viem has `readContracts` (plural) function too, it can even support multicall3. I'm happy to create a mini pr into this one if you want

Reply:
I also agree about the danger, as you never know how many tokens user has. maybe it's good as an example to have either pagination or limit set.

Regarding "api vendor", IMO it's good to show no-indexer example to devs, so people don't learn to be so reliant on indexers and centralised apis. But in real world, indexer/api is the most efficient way probably, so perhaps it's good to create both examples (or hooks). Fetching from contract with pagination or a hard limit, and another one from alchemy/reservoir etc..

Reply:
sorry ignore me. just noticed it's already merged